### PR TITLE
[HeteroGraph] Metis partition for heterograph

### PR DESCRIPTION
**Pull request by VoVAllen**
## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

File: src/kernel/binary_reduce.cc [hunk 1]
```diff
@@ -154,27 +154,6 @@ BcastInfo CalcBcastInfo(const std::string& op, NDArray lhs, NDArray rhs) {
   return ret;
 }
 
-// Function to convert an idarray to string
-std::string IdArrayToStr(IdArray arr) {
```
Comment (VoVAllen) [inline src/kernel/binary_reduce.cc:158]: Remove unused functions. Use aten::ToDebugString in the future

File: src/array/array.cc [hunk 2]
```diff
@@ -184,7 +206,10 @@ IdArray Relabel_(const std::vector<IdArray>& arrays) {
 }
 
 NDArray Concat(const std::vector<IdArray>& arrays) {
-  CHECK(arrays.size() > 1) << "Number of arrays should larger than 1";
+  // CHECK(arrays.size() > 1) << "Number of arrays should larger than 1";
```
Comment (zheng-da) [inline src/array/array.cc:209]: remove this?

File: src/array/cpu/array_op_impl.cc [hunk 3]
```diff
@@ -256,6 +259,53 @@ IdArray NonZero(BoolArray bool_arr) {
...
+IdArray Unique(IdArray arr) {
+  CHECK(arr->ndim == 1) << "Unique only supports 1D array";
+  const IdType* arr_data = static_cast<IdType*>(arr->data);
+  phmap::flat_hash_set<IdType> arr_set(arr_data, arr_data + arr->shape[0]);
+
+  IdArray ret_ndarray =
+    NewIdArray(arr_set.size(), DLContext{kDLCPU, 0}, sizeof(IdType) * 8);
+  IdType* ret_data = static_cast<IdType*>(ret_ndarray->data);
+  std::copy(arr_set.begin(), arr_set.end(), ret_data);
+  std::sort(ret_data, ret_data + arr_set.size());
```
Comment (zheng-da) [inline src/array/cpu/array_op_impl.cc:302]: why don't you sort the array and get unique values? it'll be faster than using hashtable.

Comment (VoVAllen) [reply src/array/cpu/array_op_impl.cc]: Changed to use sort now

File: src/array/array.cc (hunk 2 above)
Comment (VoVAllen) [reply src/array/array.cc]: Removed

**Comment by VoVAllen** [comment]
@zheng-da I'll make partition api into another pr

**Comment by zheng-da** [comment]
is it really necessary to create the `_CAPI_HeteroGraphGetSubgraphWithHalo` CAPI? what is the use case? the partition API won't call it.

**Pull request closed without merging**